src/ir/ExpressionAnalyzer.cpp

```diff
@@ -219,7 +219,7 @@ template<typename T> void visitImmediates(Expression* curr, T& visitor) {
       visitor.visitInt(curr->op);
       visitor.visitNonScopeName(curr->nameOperand);
     }
-    void visitRefNull(RefNull* curr) {}
+    void visitRefNull(RefNull* curr) { visitor.visitType(curr->type); }
```

Comment:
This was the culprit, leading to `(ref.null any)` and `(ref.null func)` for example to compare equal, leading to invalid code folding.